src/libfsm/trim.c

```diff
@@ -14,198 +14,332 @@
+		}
 
-			for (state_set_reset(fsm->states[s].epsilons, &state_iter); state_set_next(&state_iter, &epsilon_s); ) {
-				if (marks[epsilon_s] & MARK_PROCESSING) {
-					continue; /* already being processed */
+		for (state_set_reset(fsm->states[s_id].epsilons, &state_iter);
+		     state_set_next(&state_iter, &next); ) {
+			assert(next < state_count);
+			if (!fsm->states[next].visited) {
+				if (!queue_push(q, next)) {
+					goto cleanup;
 				}
+				fsm->states[next].visited = 1;
 
-				if (stack_i == stack_ceil) {
-					size_t nceil = 2*stack_ceil;
-					fsm_state_t *nstack = f_realloc(fsm->opt->alloc,
-					    stack, nceil * sizeof(stack[0]));
-					assert(!"unreachable");
-					if (stack == NULL) {
-						f_free(fsm->opt->alloc, stack);
-						return 0;
-					}
-					stack_ceil = nceil;
-					stack = nstack;
+				if (edges == NULL) {
+					continue;
 				}
-
-				if (!(marks[epsilon_s] & MARK_ENQUEUED)) {
-					stack[stack_i] = epsilon_s;
-					stack_i++;
-					marks[epsilon_s] |= MARK_ENQUEUED;
+				if (!save_edge(fsm->opt->alloc,
+					&edge_count, &edge_ceil, &edges,
+					s_id, next)) {
+					goto cleanup;
 				}
 			}
+		}
 
-			for (edge_set_reset(fsm->states[s].edges, &edge_iter); edge_set_next(&edge_iter, &e); ) {
-				if (marks[e.state] & MARK_PROCESSING) {
-					continue;
+		for (edge_set_reset(fsm->states[s_id].edges, &edge_iter);
+		     edge_set_next(&edge_iter, &e); ) {
+			next = e.state;
+			if (!fsm->states[next].visited) {
+				if (!queue_push(q, next)) {
+					goto cleanup;
 				}
+				fsm->states[next].visited = 1;
 
-				if (stack_i == stack_ceil) {
-					size_t nceil = 2*stack_ceil;
-					fsm_state_t *nstack = f_realloc(fsm->opt->alloc,
-					    stack, nceil * sizeof(stack[0]));
-					assert(!"unreachable");
-					if (stack == NULL) {
-						f_free(fsm->opt->alloc, stack);
-						return 0;
-					}
-					stack_ceil = nceil;
-					stack = nstack;
+				if (edges == NULL) {
+					continue;
 				}
-
-				if (!(marks[e.state] & MARK_ENQUEUED)) {
-					stack[stack_i] = e.state;
-					stack_i++;
-					marks[e.state] |= MARK_ENQUEUED;
+				if (!save_edge(fsm->opt->alloc,
+					&edge_count, &edge_ceil, &edges,
+					s_id, next)) {
+					goto cleanup;
 				}
 			}
-		} else if (!(marks[s] & MARK_DONE)) {
-			/* Second pass: Mark the state as reaching an
-			 * end if any of its reachable nodes do, mark
-			 * processing as done. Terminate iteration once
-			 * the END flag is set. */
-			for (state_set_reset(fsm->states[s].epsilons, &state_iter);
-			     state_set_next(&state_iter, &epsilon_s); ) {
-				if (marks[s] & MARK_REACHES_END) {
-					break;
-				}
-				if (marks[epsilon_s] & MARK_REACHES_END) {
-					marks[s] |= MARK_REACHES_END;
-				}
+		}
+	}
+
+	/* Only tracking reachability from start, so we're done. */
+	if (mode == FSM_TRIM_START_REACHABLE) {
+		queue_free(q);
+		return 1;
+	}
+
+	/* Clear mark for second pass. */
+	for (s_id = 0; s_id < state_count; s_id++) {
+		fsm->states[s_id].visited = 0;
+	}
```

Comment:
I was actually inclined to use a separately allocated `uint64_t` bit set and then free it, but that bit already exists for similar uses in other states. I think the protocol is that individual uses can assume it will start zeroed, but need to clear it themselves when done.